aptos-move/block-executor/src/view.rs

```diff
@@ -169,39 +160,80 @@ impl<
     }
 }
 
-enum ViewMapKind<'a, T: Transaction, X: Executable> {
-    MultiVersion(&'a MVHashMapView<'a, T::Key, T::Value, X>),
-    Unsync(&'a UnsyncMap<T::Key, T::Value, X>),
+struct SequentialState<'a, T: Transaction, X: Executable> {
+    unsync_map: &'a UnsyncMap<T::Key, T::Value, X>,
+    _counter: u32,
+}
+
+enum ViewState<'a, T: Transaction, X: Executable> {
+    Sync(ParallelState<'a, T, X>),
+    Unsync(SequentialState<'a, T, X>),
 }
 
+/// A struct that represents a single block execution worker thread's view into the state,
+/// some of which (in Sync case) might be shared with other workers / threads. By implementing
+/// all necessary traits, LatestView is provided to the VM and used to intercept the reads.
+/// In the Sync case, also records captured reads for later validation. latest_txn_idx
+/// must be set according to the latest transaction that the worker was / is executing.
 pub(crate) struct LatestView<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> {
     base_view: &'a S,
-    latest_view: ViewMapKind<'a, T, X>,
-    txn_idx: TxnIndex,
+    latest_view: ViewState<'a, T, X>,
+    latest_txn_idx: Option<TxnIndex>,
 }
 
 impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> LatestView<'a, T, S, X> {
-    pub(crate) fn new_mv_view(
+    pub(crate) fn new_sync_view(
         base_view: &'a S,
-        map: &'a MVHashMapView<'a, T::Key, T::Value, X>,
-        txn_idx: TxnIndex,
+        shared_map: &'a MVHashMap<T::Key, T::Value, X>,
+        shared_scheduler: &'a Scheduler,
+        shared_counter: &'a AtomicU32,
     ) -> LatestView<'a, T, S, X> {
         LatestView {
             base_view,
-            latest_view: ViewMapKind::MultiVersion(map),
-            txn_idx,
+            latest_view: ViewState::Sync(ParallelState::new(
+                shared_map,
+                shared_scheduler,
+                shared_counter,
+            )),
+            latest_txn_idx: None,
         }
     }
 
-    pub(crate) fn new_btree_view(
+    pub(crate) fn new_unsync_view(
         base_view: &'a S,
         map: &'a UnsyncMap<T::Key, T::Value, X>,
-        txn_idx: TxnIndex,
     ) -> LatestView<'a, T, S, X> {
         LatestView {
             base_view,
-            latest_view: ViewMapKind::Unsync(map),
-            txn_idx,
+            latest_view: ViewState::Unsync(SequentialState {
+                unsync_map: map,
+                _counter: 0,
+            }),
+            latest_txn_idx: None,
+        }
+    }
+
+    /// Records the index of the transaction that the worker is executing next. In parallel (Sync)
+    /// setting, clears captured reads to prepare for the next VM execution.
+    pub(crate) fn update_txn_idx(&mut self, txn_idx: TxnIndex) {
```

Comment:
it's ugly that txn_idx is None and has to be updated before execution every time, I'd prefer to keep the old style construct with txn idx. we can carry shared information in ViewState

Reply:
done!